packages/rum-recorder/src/domain/rrweb/mutationObserver.ts

```diff
@@ -0,0 +1,274 @@
+      addedAndMovedNodes.add(node)
+    })
+    forEach(mutation.removedNodes, (node) => {
+      if (!addedAndMovedNodes.has(node)) {
+        removedNodes.set(node, mutation.target)
+      }
+      addedAndMovedNodes.delete(node)
+    })
+  }
+
+  // Then, we sort nodes that are still in the document by topological order, for two reasons:
+  //
+  // * We will serialize each added nodes with their descendants. We don't want to serialize a node
+  // twice, so we need to iterate over the parent nodes first and skip any node that is contained in
+  // a precedent node.
+  //
+  // * To emit "add" mutations, we need references to the parent and potential next sibling of each
+  // added node. So we need to iterate over the parent nodes first, and when multiple nodes are
+  // siblings, we want to iterate from last to first. This will ensure that any "next" node is
+  // already serialized and have an id.
+  const sortedAddedAndMovedNodes = Array.from(addedAndMovedNodes)
+  sortAddedAndMovedNodes(sortedAddedAndMovedNodes)
+
+  // Then, we iterate over our sorted node sets to emit mutations. We collect the newly serialized
+  // node ids in a map to be able to skip subsequent related mutations.
+  const serializedNodesIdMap: IdNodeMap = {}
+
+  const addedNodeMutations: AddedNodeMutation[] = []
+  for (const node of sortedAddedAndMovedNodes) {
+    if (hasBeenSerialized(node)) {
+      continue
+    }
+
+    const serializedNode = serializeNodeWithId(node, { doc: document, map: serializedNodesIdMap, skipChild: false })
+    if (!serializedNode) {
+      continue
+    }
+
+    addedNodeMutations.push({
+      nextId: getNextSibling(node),
+      parentId: getSerializedNodeId(node.parentNode!)!,
+      node: serializedNode,
+    })
+  }
+
+  // Finally, we emit remove mutations.
+  const removedNodeMutations: RemovedNodeMutation[] = []
+  removedNodes.forEach((parent, node) => {
+    if (hasSerializedNode(node)) {
+      removedNodeMutations.push({
+        parentId: getSerializedNodeId(parent),
+        id: getSerializedNodeId(node),
+      })
+    }
+  })
+
+  return { adds: addedNodeMutations, removes: removedNodeMutations, hasBeenSerialized }
+
+  function hasBeenSerialized(node: Node) {
+    return hasSerializedNode(node) && serializedNodesIdMap.hasOwnProperty(getSerializedNodeId(node))
+  }
+
+  function getNextSibling(node: Node): null | number {
+    let nextSibling = node.nextSibling
+    while (nextSibling) {
+      if (hasSerializedNode(nextSibling) && !nodeIsIgnored(nextSibling)) {
+        return getSerializedNodeId(nextSibling)
+      }
+      nextSibling = nextSibling.nextSibling
+    }
+
+    return null
+  }
+}
+
+function processCharacterDataMutations(mutations: Array<WithSerializedTarget<CharacterDataMutationRecord>>) {
+  const textMutations: TextMutation[] = []
+
+  // Deduplicate mutations based on their target node
+  const handledNodes = new Set<Node>()
+  const filteredMutations = mutations.filter((mutation) => {
```

Comment:
can we live with just 1 loop here? 🤔 

```
const textMutations: TextMutation[] = []
const handledNodes = new Set<Node>()

for (n in mutations) {
   if (handledNodes.has(mutation.target)) {
       continue;
    }
    handledNodes.add(mutation.target)
    const value = mutation.target.textContent
    if (value === mutation.oldValue) {
        continue
    }
    textMutations.push({
        id: getSerializedNodeId(mutation.target),
        value,
    })

```

Reply:
This was my first version for both `processCharacterDataMutations` and `processAttributesMutations`. But, for attributes, the loop got a bit hairy... so I split the loop into "deduplication" and "emission". To have a consistent pattern between the two functions, I applied the same structure to character data mutations. I find it a bit easier to read, and it shouldn't matter performance wise.